File: resources/lua/Objects/Player.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__ClassExtends = ____lualib.__TS__ClassExtends
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["7"] = 1,["8"] = 1,["9"] = 3,["10"] = 3,["11"] = 3,["12"] = 3,["14"] = 3,["15"] = 6,["16"] = 4,["17"] = 9,["18"] = 10,["19"] = 11,["20"] = 9,["21"] = 14,["22"] = 15,["23"] = 16,["25"] = 20,["26"] = 22,["28"] = 26,["29"] = 27,["30"] = 28,["32"] = 32,["34"] = 14});
local ____exports = {}
local ____Base = require("Objects.Base")
local Base = ____Base.Base
____exports.Player = __TS__Class()
local Player = ____exports.Player
Player.name = "Player"
__TS__ClassExtends(Player, Base)
function Player.prototype.____constructor(self)
    Base.prototype.____constructor(self)
    print("new player")
end
function Player.prototype.PostInit(self)
    self.x = 200
    self.y = 200
end
function Player.prototype.Update(self)
    if IsKeyDown(nil, 263) then
        self.angle = self.angle - 2
    end
    if IsKeyDown(nil, 262) then
        self.angle = self.angle + 2
    end
    if IsKeyDown(nil, 265) then
        self.x = self.x + 2 * math.cos(self.angle * math.pi / 180)
        self.y = self.y + 2 * math.sin(self.angle * math.pi / 180)
    end
    if IsKeyDown(nil, 264) then
    end
end
return ____exports

File: ts/Objects/Player.ts
import {Base} from "./Base";
import {RegisterClass, SpawnEntity} from "../Entity";
import {Shot} from "./Shot";
import {VectorDrawable} from "../Drawable";
import Vector3 = rl.Vector3;
import {Global} from "../Global";

export class Player extends Base {

    velocity: Vector3 = new Vector3();

    constructor() {
        super();
        console.log("new player")

        this.drawable = new VectorDrawable(_G.resources.player_ship);
    }

    PostInit() {
        this.x = 0;
        this.y = 0;
    }

    Update() {

        let keys = Global.playerController.keys;
        let pc = Global.playerController;

        if ((keys & 0x04) != 0) {
            this.angle += 2;
        }

        if ((keys & 0x08) != 0) {
            // Right pressed
            this.angle -= 2;
        }

        if(this.angle < 0) this.angle += 360;
        if(this.angle > 360) this.angle -= 360;

        if ((keys & 0x01) != 0) {
            this.velocity.x += pc.thrust_fac * .01 * Math.cos(this.angle * Math.PI / 180.0);
            this.velocity.y -= pc.thrust_fac * .01 * Math.sin(this.angle * Math.PI / 180.0);
            // this.y -= 2;
        }

        if (rl.IsKeyDown(264)) {
            // this.y += 2;
        }

        if(rl.IsKeyPressed(rl.KeyboardKey.KEY_SPACE)) {
            let shot: Shot = SpawnEntity(Shot);
            if(shot != null) {
                shot.x = this.x;
                shot.y = this.y;
                shot.angle = this.angle;
            }
        }

        this.velocity.x *= pc.drag_fac; //0.98;
        this.velocity.y *= pc.drag_fac; //0.98;

        this.x += this.velocity.x;
        this.y += this.velocity.y;
    }

    public ReadFromSnapshot(msgbuf: any) {
        super.ReadFromSnapshot(msgbuf);
        console.log("Player.ReadFromSnapshot", msgbuf);

        let val = MSG_ReadInt(msgbuf);
        console.log("got val: ", val);
    }

    public GetDrawable(): any {
       return _G.resources.player_ship;
    }
}

RegisterClass(Player);